Sprint 15

Hajime4life opened the pull request.

Hajime4life committed:
Добавления / Изменения:
- Изменена структура проекта с перераспределением файлов проекта
- Интегрирована интерактивная панель для выбора эмодзи и цветов при настройке трекера
- Настроено и добавлено хранилище данных
- Введен набор кодов ошибок для обработки проблем с декодированием данных
- Добавлены записи логов для анализа ошибок
- Разработан процесс сохранения трекера с учетом возможных ошибок

Hajime4life committed:
Расширена функциональность интерфейса для создания нерегулярных событий

Helvetii951 commented on Tracker/App/AppDelegate.swift:
```diff
@@ -13,4 +14,34 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        return container
+    }()
+
+    static var viewContext: NSManagedObjectContext {
+        guard let delegate = UIApplication.shared.delegate as? AppDelegate else {
+            fatalError("Unable to get AppDelegate for Core Data")
+        }
+        return delegate.persistentContainer.viewContext
+    }
```
Сейчас в приложении используется контекст, связанный с главным потоком.

В CoreData предусмотрена многопоточность:

[Доки Apple](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CoreData/Concurrency.html)
[Неплохая вводная статья](https://cocoacasts.com/core-data-and-concurrency)

Все хорошо, просто потом почитай.

Helvetii951 commented on Tracker/App/AppDelegate.swift:
```diff
@@ -13,4 +14,34 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        guard context.hasChanges else { return }
+        do {
+            try context.save()
+        } catch {
+            let nserror = error as NSError
+            fatalError("Context save error: \(nserror), \(nserror.userInfo)")
+        }
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Рекомендую использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfaxrilure(_:file:line:)).
Часто в таких случаях пишут ошибки в логи или сервисы статистики и событий.

Helvetii951 commented on Tracker/Controllers/Calendar/CalendarViewController.swift:
```diff
@@ -0,0 +1,71 @@
+        }
+        picker.locale = Locale(identifier: "ru_RU")
+        picker.translatesAutoresizingMaskIntoConstraints = false
+        picker.addTarget(self, action: #selector(datePickerChanged(_:)), for: .valueChanged)
```
Лучше не привзваться к конкретной локали. Пусть отображение даты будет отличаться от дизайна, но пользователь увидит дату в привычном для себя виде.

Helvetii951 commented on Tracker/Controllers/Calendar/CalendarViewController.swift:
Этот файл не включен ни в один таргет. Лучше удалить.

В проекте используется [этот файл](https://github.com/Hajime4life/Tracker/blob/sprint_15/Tracker/Controllers/Tracker/CalendarViewController.swift).

Helvetii951 commented on Tracker/CoreData/TrackerStore.swift:
```diff
@@ -0,0 +1,185 @@
+final class TrackerStore: NSObject {
+    // MARK: - Properties
+    weak var delegate: TrackerStoreDelegate?
+    
```
`import UIKit` можно убрать. Сторам очень желательно ничего не знать о UI слое.

Hajime4life committed:
Исправления:
- Поправил мультиселекцию (теперь можно выбирать элементы в понели по ТЗ)
- Исправлена последовательность эмодзи

Hajime4life committed:
Удален старый календарь (не используется)

Hajime4life committed:
Исправлен баг с исчезновением выделения после скролла

Hajime4life committed:
Убрал серый фон у секций с эмоджи и цветами

Hajime4life committed:
Сделать бордер у выбранного цвета прозрачным в соответствии с макетом

Hajime4life committed:
Починил отображение треков

Hajime4life committed:
Исправлена рассинхронизаци ID при сохранении трекера

Hajime4life committed:
Исправления:
- Поправил отсупы от кнопки "Расписание"
- Увеличил размери емоджи по размеру указаному в макете
